### PR TITLE
Tensor fusion for allgather

### PR DESCRIPTION
**Pull request by abditag2**
Enables Tensor fusion for allgather operation

### CODE REVIEW

File: horovod/common/mpi_message.h [hunk 1]
```diff
@@ -141,6 +141,9 @@ class MPIResponse {
   const std::vector<int64_t>& tensor_sizes() const;
   void set_tensor_sizes(const std::vector<int64_t>& value);
   void add_tensor_sizes(int64_t value);
+  // To fuse multiple allgater responses
```
Comment (alsrgv) [inline horovod/common/mpi_message.h]: Typo

File: horovod/common/operations.cc [hunk 2]
```diff
@@ -1976,11 +2100,12 @@ bool RunLoopOnce(HorovodGlobalState& state, bool is_coordinator) {
 
     MPIResponseList response_list;
     response_list.set_shutdown(should_shut_down);
-
+//    std::this_thread::sleep_for(std::chrono::seconds(1));
```
Comment (alsrgv) [inline horovod/common/operations.cc]: Nuke?

File: horovod/common/operations.cc [hunk 3]
```diff
@@ -2013,7 +2138,71 @@ bool RunLoopOnce(HorovodGlobalState& state, bool is_coordinator) {
...
+            for (int i = 1; i < new_entry.tensor->shape().dims(); ++i) {
+              new_total_count_of_output_entries *= new_entry.tensor->shape().dim_size(i);
+            }
+
+            int new_element_size;
+            MPI_Type_size(GetMPIDataType(new_entry.tensor), &new_element_size);
+            int64_t new_total_byte_size_of_output =
+                new_total_count_of_output_entries * new_element_size;
+
+              if (response.response_type() == new_response.response_type() &&
```
Comment (alsrgv) [inline horovod/common/operations.cc]: Can you re-format using `clang-format`?  Also, we typically don't add space in the beginning of the loop, etc.

File: horovod/common/operations.cc [hunk 4]
```diff
@@ -808,72 +810,107 @@ void PerformOperation(TensorTable& tensor_table, MPIResponse response) {
...
+              entry_component_offsets[ec - 1][rc] +
+              entry_component_sizes[ec - 1][rc];
+        }
       }
+      rank_displacement += recvcounts[rc];
     }
 
     int element_size;
-    MPI_Type_size(GetMPIDataType(e.tensor), &element_size);
+    MPI_Type_size(GetMPIDataType(entries[0].tensor), &element_size);
```
Comment (alsrgv) [inline horovod/common/operations.cc]: Use `first_entry` instead of `entries[0]`?

File: horovod/common/operations.cc [hunk 5]
```diff
@@ -916,58 +953,137 @@ void PerformOperation(TensorTable& tensor_table, MPIResponse response) {
...
-      // Copy the result from MPI shared memory to rank-specific output buffer
-      ACTIVITY_START_ALL(entries, timeline, COPY_ALLGATHER_OUTPUT)
-      memcpy((void*)e.output->data(), horovod_global.shared_buffer,
-             total_size * element_size);
+      // Copy memory out of the fusion buffer.
+      ACTIVITY_START_ALL(entries, timeline, MEMCPY_OUT_FUSION_BUFFER)
+      for (int ec = 0; ec < entries.size(); ec++) {
+        auto& e = entries[ec];
+        int64_t copy_offset = 0;
+        for (int rc = 0; rc < num_participating_ranks; rc++) {
```
Comment (alsrgv) [inline horovod/common/operations.cc]: Do we need to make it `O(# ranks)`?  Could we make it `O(|entries|)`?

File: horovod/common/mpi_message.cc [hunk 6]
```diff
@@ -285,6 +285,21 @@ void MPIResponse::add_tensor_sizes(int64_t value) {
...
+    horovod::common::MPIResponse response) {
+  assert(response_type() == MPIResponse::ResponseType::ALLGATHER);
+  assert(response.tensor_names().size() == 1);
+  assert(response.devices() == devices());
+  assert(response.tensor_sizes().size() == tensor_sizes().size());
+  add_tensor_names(response.tensor_names()[0]);
+  add_tensor_sizes(response.tensor_sizes()[0]);
+}
+
+unsigned int MPIResponse::num_participating_ranks() const {
```
[the remark this answers is not in the document]
Comment (alsrgv) [reply horovod/common/mpi_message.cc]: Let's revise per our last discussion

File: horovod/common/operations.cc (hunk 5 above)
Comment (alsrgv) [reply horovod/common/operations.cc]: Nevermind. :-)

File: test/test_tensorflow.py [hunk 7]
```diff
@@ -383,6 +383,102 @@ def test_horovod_allgather(self):
...
+                    if dtype != tf.bool:
+                        value = i
+                    else:
+                        value = i % 2
+
+                    tests.append(tf.reduce_all(
+                        tf.equal(tf.cast(rank_tensor, tf.int32), value)))
+
+            self.assertTrue(session.run(tf.reduce_all(tests)),
+                            "hvd.allgather produces incorrect gathered tensor")
```
Comment (abditag2) [inline test/test_tensorflow.py]: @alsrgv the output of tf.equal has three different shapes. we need to do reduce_all twice to get to make sure all the comparisons are True.

File: test/test_tensorflow.py [hunk 8]
```diff
@@ -383,6 +383,101 @@ def test_horovod_allgather(self):
...
+                    if dtype != tf.bool:
+                        value = i
+                    else:
+                        value = i % 2
+
+                    tests.append(tf.reduce_all(
+                        tf.equal(tf.cast(rank_tensor, tf.int32), value)))
+
+            self.assertTrue(
+                session.run(tests),
```
[the remark this answers is not in the document]
Comment (abditag2) [reply test/test_tensorflow.py]: Updated. Please check my comment.

File: test/test_tensorflow.py [hunk 9]
```diff
@@ -383,6 +383,102 @@ def test_horovod_allgather(self):
...
+            dims = [1, 2, 3]
+            tests = []
+            for dtype, dim in itertools.product(dtypes, dims):
+                tensor = tf.ones([17] * dim) * rank
+                if dtype == tf.bool:
+                    tensor = tensor % 2
+                tensor = tf.cast(tensor, dtype=dtype)
+                gathered = hvd.allgather(tensor)
+
+                gathered_tensor = session.run(gathered)
```
Comment (alsrgv) [inline test/test_tensorflow.py]: Same as the comment below.

File: horovod/common/operations.cc [hunk 10]
```diff
@@ -808,72 +810,108 @@ void PerformOperation(TensorTable& tensor_table, MPIResponse response) {
...
+      TensorShape single_slice_shape;
+      for (int i = 1; i < e.tensor->shape().dims(); ++i) {
+        single_slice_shape.AddDim(e.tensor->shape().dim_size(i));
+      }
+
+      // Copy tensor sizes from the MPI response into a vector of int64_t
+      // and compute total size.  This is size of first dimension.
+      int64_t total_entry_dimension_size = 0;
+      for (unsigned int rc = 0; rc < horovod_global.size; ++rc) {
+        tensor_sizes.push_back(
```
Comment (alsrgv) [inline horovod/common/operations.cc]: `tensor_sizes` is not used anymore with your code change, we can nuke it.

File: horovod/common/operations.cc [hunk 11]
```diff
@@ -808,72 +810,108 @@ void PerformOperation(TensorTable& tensor_table, MPIResponse response) {
...
+            response.tensor_sizes()[ec * horovod_global.size + rc]);
+        total_entry_dimension_size +=
```
Comment (alsrgv) [inline horovod/common/operations.cc]: Introduce local variable for `response.tensor_sizes()[ec * horovod_global.size + rc]` ?

File: horovod/common/operations.cc [hunk 12]
```diff
@@ -1993,27 +2108,95 @@ bool RunLoopOnce(HorovodGlobalState& state, bool is_coordinator) {
...
+              total_count_of_output_entries * element_size;
+
+          while (!responses.empty()) {
+
+            auto new_response = responses.front();
+            assert(new_response.tensor_names().size() == 1);
+            auto& new_entry =
+                state.tensor_table[new_response.tensor_names()[0]];
+
+            int64_t new_total_dimension_size = 0;
```
Comment (alsrgv) [inline horovod/common/operations.cc]: Let's make a helper function that computes this element # for a `tensor_sizes` list and an entry (or a tensor) - to simplify the code and DRY.  Is there any other place we could use that function?

File: horovod/common/operations.cc [hunk 13]
```diff
@@ -1993,27 +2113,92 @@ bool RunLoopOnce(HorovodGlobalState& state, bool is_coordinator) {
...
+              total_count_of_output_entries * element_size;
+
+          while (!responses.empty()) {
+
+            auto new_response = responses.front();
+            assert(new_response.tensor_names().size() == 1);
+            auto& new_entry =
+                state.tensor_table[new_response.tensor_names()[0]];
+
+            int64_t new_total_dimension_size =
```
Comment (alsrgv) [inline horovod/common/operations.cc]: Can you refactor it to return `total_count_of_output_entries`?  That's the piece of computation you're duplicating as well.

File: test/test_tensorflow.py [hunk 14]
```diff
@@ -383,6 +383,118 @@ def test_horovod_allgather(self):
...
+                        value = i % 2
+
+                    tests.append(
+                        tf.reduce_all(
+                            tf.equal(tf.cast(rank_tensor, tf.int32), value)))
+
+            gathered_tensors = session.run(tensors_to_gather)
+            gathered_tensors_it = iter(gathered_tensors)
+            for dtype, dim in itertools.product(dtypes, dims):
+                tensor_shape, expected_shape = list(next(gathered_tensors_it).shape), [17 * size] + [17] * (dim - 1)
```
Comment (alsrgv) [inline test/test_tensorflow.py]: You can use `tf.shape()` and make this condition a TensorFlow Op, that you could add to the list of ops in the loop above.  That way, here you could just do `self.assertTrue(session.run(tf.reduce_all(shape_tests)), ...)`

File: horovod/common/mpi_message.cc [hunk 15]
```diff
@@ -285,8 +285,20 @@ void MPIResponse::add_tensor_sizes(int64_t value) {
   tensor_sizes_.push_back(value);
 }
 
+void MPIResponse::add_allgather_response(
+    horovod::common::MPIResponse response) {
+  assert(response_type() == MPIResponse::ResponseType::ALLGATHER);
+  assert(response.tensor_names().size() == 1);
+  assert(response.devices() == devices());
+  assert(response.tensor_sizes().size() == tensor_sizes().size());
```
Comment (alsrgv) [inline horovod/common/mpi_message.cc]: Can you assert that?  What if you fuse three allgather ops together?

File: horovod/common/operations.cc [hunk 16]
```diff
@@ -1993,27 +2135,65 @@ bool RunLoopOnce(HorovodGlobalState& state, bool is_coordinator) {
...
 
+          while (!responses.empty()) {
+
+            auto new_response = responses.front();
+            assert(new_response.tensor_names().size() == 1);
+            auto& new_entry =
+                state.tensor_table[new_response.tensor_names()[0]];
+
+            int64_t new_total_byte_size_of_output =
+                TotalByteSizeOfOutput(new_response.tensor_sizes(), new_entry);
```
Comment (alsrgv) [inline horovod/common/operations.cc]: Nit: rename to TotalByteSizeOfAllgatherOutput?

File: test/test_tensorflow.py [hunk 17]
```diff
@@ -383,6 +383,114 @@ def test_horovod_allgather(self):
...
                         "hvd.allgather produces incorrect gathered tensor")
 
+    def test_horovod_allgather_fused(self):
+        """Test that the allgather correctly gathers 1D, 2D, 3D tensors
+        with Tensor Fusion."""
+        hvd.init()
+        rank = hvd.rank()
+        size = hvd.size()
+
+        with self.test_session(config=self.config) as session:
```
Comment (alsrgv) [inline test/test_tensorflow.py]: Sorry, can you fix tests as well?  With #670 there's no `self.config` anymore, just `config`.

File: test/test_tensorflow.py [hunk 18]
```diff
@@ -383,6 +383,114 @@ def test_horovod_allgather(self):
...
+
+                shape_tests.append(
+                    tf.reduce_all(tf.equal(tf.shape(gathered),
+                                           [17 * size] + [17] * (dim - 1))))
+
+                for i in range(size):
+                    rank_tensor = tf.slice(gathered,
+                                           [i * 17] + [0] * (dim - 1),
+                                           [17] + [-1] * (dim - 1))
+                    # self.assertEqual(list(rank_tensor.shape), [17] * dim)
```
Comment (alsrgv) [inline test/test_tensorflow.py]: Remove self.assertEqual() comment, move tf.equal comment closer to where you do the cast?

File: test/test_tensorflow.py [hunk 19]
```diff
@@ -383,6 +383,114 @@ def test_horovod_allgather(self):
...
+                    tf.reduce_all(tf.equal(tf.shape(gathered),
+                                 [sum(tensor_sizes)] + [17] * (dim - 1))))
+
+                for i in range(size):
+                    rank_size = [tensor_sizes[i]] + [17] * (dim - 1)
+                    rank_tensor = tf.slice(
+                        gathered, [sum(tensor_sizes[:i])] + [0] * (dim - 1),
+                        rank_size)
+                    self.assertEqual(list(rank_tensor.shape), rank_size)
+                    # tf.equal() does not support tf.uint16 as of TensorFlow 1.2,
```
Comment (alsrgv) [inline test/test_tensorflow.py]: Move comment closer to the cast?

**Pull request merged**